File: demo-android/app/src/main/java/com/example/demo/ui/tasklist/TaskListViewModel.kt
package com.example.demo.ui.tasklist

import android.app.Application
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.ViewModel
import androidx.lifecycle.ViewModelProvider
import androidx.lifecycle.viewModelScope
import com.example.demo.data.model.Task
import com.example.demo.repository.TaskRepository
import com.example.demo.ui.i18n.mapTaskError
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch

data class TaskListUiState(
    val tasks: List<Task> = emptyList(),
    val isLoading: Boolean = true,
    val errorMessage: String? = null,
    val deletingTaskIds: Set<String> = emptySet()
)

class TaskListViewModel(
    application: Application,
    private val repository: TaskRepository
) : AndroidViewModel(application) {

    private val _uiState = MutableStateFlow(TaskListUiState())
    val uiState: StateFlow<TaskListUiState> = _uiState.asStateFlow()

    init {
        loadTasks()
    }

    fun loadTasks() {
        viewModelScope.launch {
            _uiState.update { it.copy(isLoading = true, errorMessage = null) }
            runCatching { repository.getTasks() }
                .onSuccess { tasks ->
                    _uiState.update { it.copy(tasks = tasks, isLoading = false) }
                }
                .onFailure { error ->
                    _uiState.update {
                        it.copy(
                            tasks = emptyList(),
                            isLoading = false,
                            errorMessage = mapTaskError(getApplication(), error)
                        )
                    }
                }
        }
    }

    fun deleteTask(task: Task) {
        viewModelScope.launch {
            _uiState.update { it.copy(deletingTaskIds = it.deletingTaskIds + task.id) }
            runCatching { repository.deleteTask(task.id) }
                .onSuccess {
                    _uiState.update {
                        it.copy(
                            tasks = it.tasks.filter { item -> item.id != task.id },
                            deletingTaskIds = it.deletingTaskIds - task.id
                        )
                    }
                }
                .onFailure { error ->
                    _uiState.update {
                        it.copy(
                            deletingTaskIds = it.deletingTaskIds - task.id,
                            errorMessage = mapTaskError(getApplication(), error)
                        )
                    }
                }
        }
    }

    fun clearError() {
        _uiState.update { it.copy(errorMessage = null) }
    }

    class Factory(
        private val application: Application,
        private val repository: TaskRepository
    ) : ViewModelProvider.Factory {
        @Suppress("UNCHECKED_CAST")
        override fun <T : ViewModel> create(modelClass: Class<T>): T {
            return TaskListViewModel(application, repository) as T
        }
    }
}

File: demo-ios/Demo/UI/TaskList/TaskListViewModel.swift
import Foundation
import Observation

struct TaskListUiState: Equatable {
    var tasks: [Task] = []
    var isLoading = true
    var isRefreshing = false
    var errorMessage: String?
    var snackbarMessage: String?
    var deletingTaskIds: Set<String> = []
}

@Observable
@MainActor
final class TaskListViewModel {
    private(set) var uiState = TaskListUiState()

    private let repository: TaskRepositoryProtocol
    private let locale: AppLocale

    init(repository: TaskRepositoryProtocol, locale: AppLocale = .shared, loadsTasksOnInit: Bool = true) {
        self.repository = repository
        self.locale = locale
        if loadsTasksOnInit {
            _Concurrency.Task { await loadTasks() }
        } else {
            uiState.isLoading = false
        }
    }

    func loadTasks() async {
        uiState.isLoading = true
        uiState.errorMessage = nil
        do {
            let tasks = try await repository.getTasks()
            uiState.tasks = tasks
            uiState.isLoading = false
        } catch {
            uiState.tasks = []
            uiState.isLoading = false
            uiState.errorMessage = ErrorMessages.mapTaskError(error, locale: locale)
        }
    }

    func refreshTasks() async {
        guard !uiState.isRefreshing else { return }
        uiState.isRefreshing = true
        uiState.errorMessage = nil
        uiState.snackbarMessage = nil
        do {
            uiState.tasks = try await repository.getTasks()
            uiState.isRefreshing = false
        } catch {
            uiState.isRefreshing = false
            let message = ErrorMessages.mapTaskError(error, locale: locale)
            uiState.errorMessage = message
            uiState.snackbarMessage = message
        }
    }

    func deleteTask(_ task: Task) async {
        uiState.deletingTaskIds.insert(task.id)
        do {
            try await repository.deleteTask(id: task.id)
            uiState.tasks.removeAll { $0.id == task.id }
            uiState.deletingTaskIds.remove(task.id)
        } catch {
            uiState.deletingTaskIds.remove(task.id)
            let message = ErrorMessages.mapTaskError(error, locale: locale)
            uiState.errorMessage = message
            uiState.snackbarMessage = message
        }
    }

    func clearError() {
        uiState.errorMessage = nil
    }

    func clearSnackbar() {
        uiState.snackbarMessage = nil
    }
}
